client/extensions/woocommerce/state/sites/product-categories/selectors.js

```diff
@@ -4,31 +4,123 @@
+ */
+export const areProductCategoriesLoaded = ( state, query, siteId = getSelectedSiteId( state ) ) => {
+	const serializedQuery = getSerializedProductCategoriesQuery( query );
+	const cats = get(
+		state,
+		[
+			'extensions',
+			'woocommerce',
+			'sites',
+			siteId,
+			'productCategories',
+			'queries',
+			serializedQuery,
+		],
+		false
+	);
+	return isArray( cats );
+};
+
+/**
+ * @param {Object} state Whole Redux state tree
+ * @param {Object} [query] Query used to fetch product categories. If not provided, API defaults are used.
+ * @param {Number} [siteId] Site ID to check. If not provided, the Site ID selected in the UI will be used
+ * @return {boolean} Whether product categories are currently being retrieved from the server
+ */
+export const areProductCategoriesLoading = (
+	state,
+	query = {},
+	siteId = getSelectedSiteId( state )
+) => {
+	const serializedQuery = getSerializedProductCategoriesQuery( query );
+	const isLoading = get( state, [
+		'extensions',
+		'woocommerce',
+		'sites',
+		siteId,
+		'productCategories',
+		'isQueryLoading',
+		serializedQuery,
+	] );
+	// Strict check because it could also be undefined.
+	return true === isLoading;
+};
 
 /**
  * Gets product categories from API data.
  *
  * @param {Object} rootState Global state tree
- * @param {Number} [ siteId ] wpcom site id, if not provided, uses the selected site id.
+ * @param {Object} [query] Query used to fetch product categories. If not provided, API defaults are used.
+ * @param {Number} [siteId] wpcom site id, if not provided, uses the selected site id.
  * @return {Array} List of product categories
  */
-export function getProductCategories( rootState, siteId = getSelectedSiteId( rootState ) ) {
-	const categories = get(
+export const getProductCategories = (
+	rootState,
+	query = {},
+	siteId = getSelectedSiteId( rootState )
+) => {
+	if ( ! areProductCategoriesLoaded( rootState, query, siteId ) ) {
+		return [];
+	}
+	const serializedQuery = getSerializedProductCategoriesQuery( query );
+	const cats = get(
 		rootState,
-		[ 'extensions', 'woocommerce', 'sites', siteId, 'productCategories' ],
+		[ 'extensions', 'woocommerce', 'sites', siteId, 'productCategories', 'items' ],
+		{}
+	);
+	const idsForQuery = get(
+		rootState,
+		[
+			'extensions',
+			'woocommerce',
+			'sites',
+			siteId,
+			'productCategories',
+			'queries',
+			serializedQuery,
+		],
 		[]
 	);
-	if ( ! isArray( categories ) ) {
-		return [];
+
+	if ( idsForQuery.length ) {
+		return idsForQuery.map( id => cats[ id ] );
 	}
-	return categories;
-}
+	return false;
+};
+
+/**
+ * @param {Object} state Whole Redux state tree
+ * @param {Object} [query] Query used to fetch product categories. If not provided, API defaults are used.
+ * @param {Number} [siteId] Site ID to check. If not provided, the Site ID selected in the UI will be used
+ * @return {Number} Total number of product categories available for a query, or 0 if not loaded yet.
```

Comment:
Should we return 0 by default... just wondering if a site indeed has no product categories defined, this will still return 0 even when categories are indeed loaded. Not sure if it is important or not, but will happen.

Reply:
Hmm thats a good point. This however is consistent with what we are doing for similar selectors across orders, products, etc. I think it's fine to return 0 since usage of these are usually behind something like `areProductCategoriesLoaded` so you know if you actually care about the total or not.